src/endpoints/sources.js

```diff
@@ -292,6 +319,27 @@ function pathParts(path = "") {
 }
 
 
+/**
+ * Generate an Express endpoint that sends a group overview
+ * determined by the request.
+ *
+ * @param {sourceExtractor} sourceExtractor - Function to provide the Source instance from the request
+ * @returns {expressEndpointAsync}
+ */
+function sendGroupOverview(sourceExtractor) {
+  return async (req, res) => {
+    const groupSource = sourceExtractor(req);
+
+    authz.assertAuthorized(req.user, authz.actions.Read, groupSource);
+
```

Comment:
I thought it would be good to add tests in authz.test.js like:

https://github.com/nextstrain/nextstrain.org/blob/0afb7b4235fd2b8575e51a25e951022bec77bd78/test/authz.test.js#L98-L102

but realized those tests only work on the dataset/narrative/group objects. This line asserts authorized access on `groupSource`, which is tested already but not with the intended effect of covering path differences (e.g. `/groups/test-private` vs. `groups/test-private/overview`).

[requests.test.js](https://github.com/nextstrain/nextstrain.org/blob/0afb7b4235fd2b8575e51a25e951022bec77bd78/test/requests.test.js) so far is testing requests without auth. It would be good to have some functional tests for requests involving auth but I'm not sure how to best implement. See PR description for manual testing which I'd like to automate.

Reply:
Nod. The authz.test.js file tests the authz system/framework itself is functioning as expected, but not that authz has been applied correctly to our request handlers. That would be, as you suggest, the job of requests.test.js (or a subsidiary).

Agreed it would be good to test authz of the actual handlers, e.g. using equivalents of what you're manually testing! I punted on this at the time because I didn't want to (purely due to dev time) extend the authn system to support test/dev-only fake authn… but that is ultimately the way to implement this.

Reply:
@tsibley

> extend the authn system to support test/dev-only fake authn

I want to do this, but not sure where to start or what's appropriate. I've looked over the relevant pages in [the dev docs](https://docs.nextstrain.org/projects/nextstrain-dot-org/en/latest/index.html). A few questions:

1. Does this call for a new strategy for `passport.authenticate()`, something like [passport-mock-strategy](https://github.com/cszatmary/passport-mock-strategy)?
2. If these tests are added in requests.test.js, do we need to send requests with header `Authorization: Bearer <token>`, like [how `Accept: <mediatype>` is currently sent](https://github.com/nextstrain/nextstrain.org/blob/98e6c6c186edc0f867f81cfb30da7e7c88a76652/test/requests.test.js#L305-L311)? Or would a mock strategy from (1) eliminate the need for this?
3. Would it be appropriate to create "test users" in the `nextstrain.org` Cognito User Pool under `test`/`test-private` groups, with varying role membership? If so, what's the proper way to authenticate as these users?

Reply:
1. I don't think a new Passport strategy is necessary, because…
2. …we can modify what JWTs [our Bearer auth accepts](https://github.com/nextstrain/nextstrain.org/blob/98e6c6c186edc0f867f81cfb30da7e7c88a76652/src/authn/index.js#L133-L146) in dev/test mode and then pass something acceptable in the test requests. Very important to make sure the dev/test code paths can't be enabled in production.
3. I don't think this will be necessary now (we can just pretend users exist), although it may be in the future. In either case, we would authenticate as them by the tests offering a fake JWT, not by interacting with Cognito.

FWIW, while automated tests are good here, I don't think we need to block implementation or deploy on having tests sorted out.